File: julia/intro-to-julia-solution/benchmark.py
import random
import time

Alive = True
Dead = False
directions = [(-1, -1), (-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0), (1, 1)]

def generate(width, height) -> list[list[bool]]:
    return [
        [Dead,  Alive, Dead,  Dead,  Dead,  Dead,  Dead,  Dead,  Dead,  Dead],
        [Dead,  Alive, Dead,  Dead,  Dead,  Dead,  Dead,  Alive, Alive, Alive],
        [Dead,  Alive, Dead,  Dead,  Dead,  Dead,  Dead,  Dead,  Dead,  Dead],
        [Dead,  Alive, Dead,  Alive, Dead,  Dead,  Dead,  Dead,  Alive, Alive],
        [Dead,  Dead,  Dead,  Dead,  Alive, Dead,  Dead,  Dead,  Alive, Alive],
        [Dead,  Dead,  Alive, Alive, Alive, Dead,  Dead,  Dead,  Dead,  Dead],
        [Dead,  Dead,  Dead,  Dead,  Dead,  Dead,  Dead,  Dead,  Dead,  Dead],
        [Dead,  Alive, Dead,  Dead,  Dead,  Dead,  Dead,  Alive, Dead,  Alive],
        [Alive, Dead,  Alive, Dead,  Dead,  Alive, Dead,  Dead,  Alive, Dead],
        [Dead,  Alive, Dead,  Dead,  Alive, Dead,  Alive, Dead,  Dead,  Dead]
    ]

def update(state) -> list[list[bool]]:
    def count_neighbors(state, x, y):
        count = 0
        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            if 0 <= nx < len(state) and 0 <= ny < len(state[0]):
                count += state[nx][ny]
        return count

    new_state = [[Dead for _ in range(len(state[0]))] for _ in range(len(state))]
    for x in range(len(state)):
        for y in range(len(state[0])):
            neighbors = count_neighbors(state, x, y)
            if state[x][y] == Alive:
                if neighbors < 2 or neighbors > 3:
                    new_state[x][y] = Dead
                else:
                    new_state[x][y] = Alive
            else:
                if neighbors == 3:
                    new_state[x][y] = Alive
    return new_state

start = time.time()
iterations = 1000
state = generate(50, 50)
for i in range(iterations):
    state = update(state)
print(f"  {time.time() - start} seconds")

File: julia/intro-to-julia/benchmark.jl
Alive :: Bool = true
Dead :: Bool = false

function generate(width :: UInt16, height :: UInt16):: Matrix{Bool}
    return Bool[
        Dead  Alive Dead  Dead  Dead  Dead  Dead  Dead  Dead  Dead;
        Dead  Alive Dead  Dead  Dead  Dead  Dead  Alive Alive Alive;
        Dead  Alive Dead  Dead  Dead  Dead  Dead  Dead  Dead  Dead;
        Dead  Alive Dead  Alive Dead  Dead  Dead  Dead  Alive Alive;
        Dead  Dead  Dead  Dead  Alive Dead  Dead  Dead  Alive Alive;
        Dead  Dead  Alive Alive Alive Dead  Dead  Dead  Dead  Dead;
        Dead  Dead  Dead  Dead  Dead  Dead  Dead  Dead  Dead  Dead;
        Dead  Alive Dead  Dead  Dead  Dead  Dead  Alive Dead  Alive;
        Alive Dead  Alive Dead  Dead  Alive Dead  Dead  Alive Dead;
        Dead  Alive Dead  Dead  Alive Dead  Alive Dead  Dead  Dead
    ]
end

function update(state :: Matrix{Bool}):: Matrix{Bool}
    new_state = copy(state)
    for i in 1:size(state, 1)
        for j in 1:size(state, 2)
            alive_neighbors = 0
            for x in max(1, i-1):min(size(state, 1), i+1)
                for y in max(1, j-1):min(size(state, 2), j+1)
                    if (x != i || y != j) && state[x, y] == Alive
                        alive_neighbors += 1
                    end
                end
            end
            if state[i, j] == Alive
                if alive_neighbors < 2 || alive_neighbors > 3
                    new_state[i, j] = Dead
                end
            else
                if alive_neighbors == 3
                    new_state[i, j] = Alive
                end
            end
        end
    end
    return new_state
end

@time begin
    iterations = 1000
    global state = generate(UInt16(50), UInt16(50))
    for i = 0:iterations
        global state = update(state)
    end
end